ast/src/parsed/display.rs

```diff
@@ -36,23 +36,35 @@ impl Display for ModuleStatement {
             ModuleStatement::SymbolDefinition(SymbolDefinition { name, value }) => match value {
                 SymbolValue::Machine(
                     m @ Machine {
-                        arguments:
-                            MachineArguments {
+                        arguments: MachineArguments(args),
+                        properties:
+                            MachineProperties {
+                                degree,
                                 latch,
                                 operation_id,
+                                call_selectors,
                             },
                         ..
                     },
                 ) => {
-                    if let (None, None) = (latch, operation_id) {
-                        write!(f, "machine {name} {m}")
-                    } else {
-                        write!(
-                            f,
-                            "machine {name}({}, {}) {m}",
-                            latch.as_deref().unwrap_or("_"),
-                            operation_id.as_deref().unwrap_or("_"),
+                    let args = args.iter().join(", ");
+                    let props = degree
+                        .as_ref()
+                        .map(|s| format!("degree: {s}"))
+                        .into_iter()
+                        .chain(latch.as_ref().map(|s| format!("latch: {s}")))
+                        .chain(operation_id.as_ref().map(|s| format!("operation_id: {s}")))
+                        .chain(
+                            call_selectors
+                                .as_ref()
+                                .map(|s| format!("call_selectors: {s}")),
                         )
+                        .join(", ");
+                    match (args.is_empty(), props.is_empty()) {
+                        (true, true) => write!(f, "machine {name} {m}"),
+                        (true, false) => write!(f, "machine {name} with {props} {m}"),
+                        (false, true) => write!(f, "machine {name}({args}) {m}"),
+                        (false, false) => write!(f, "machine {name}({args}) with {props} {m}"),
                     }
```

Comment:
Matter of style, but I would prefer
```suggestion
                    let props = if props.is_empty() { props } else { format!(" with {props}")) };
                    let args = if args.is_empty() { args } else { format!("({args})")) };
                    write!(f, "machine {name}{args}{props} {m}")
```